Support new aggregation operators in builder

alcaeus opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues |

#### Summary

This PR adds the following new aggregation pipeline operators to the builder: `$accumulator`, `$binarySize`, `$bottom`, `$bottomN`, `$bsonSize`, `$dateAdd`, `$dateDiff`, `$dateFromParts`, `$dateFromString`, `$dateSubtract`, `$dateToParts`, `$dateTrunc`, `$firstN`, `$function`, `$getField`, `$lastsN`, `$maxN`, `$mergeObjects`, `$minN`, `$rand`, `$regexFind`, `$regexFindAll`, `$regexMatch`, `$replaceAll`, `$replaceOne`, `$sampleRate`, `$setField`, `$sortArray`, `$top`, `$topN`, `$tsIncrement`, `$tsSecond`.

This covers all aggregation pipeline operators currently supported in MongoDB, with the exception of accumulator operators only supported in the `$setWindowFields` stage, which will be introduced in a separate pull request.

I've also taken liberty to organise the code a bit. We now have interfaces for the various operator groups instead of one generic operator interface, and I've cleaned up the logic around accumulator stages (such as `$sum`) which have two different forms depending on the stage they're being used in. In the `$group`, `$bucket`, and `$bucketAuto` stages they take a single argument, whereas they take multiple arguments in other stages such as `$project`. This is now properly accounted for and easier to handle due to a trait that handles this more gracefully.

Note: GitHub Copilot was instrumental in getting this work done as it automated away certain repetitive tasks (such as adding proxy methods to the `Operator` class. I've also marked that class as `@internal` as we may want to do away with it in the future in favour of multiple traits that add specific operator groups, similar to how we have interfaces for the groups.

alcaeus committed:
Add interfaces for aggregation pipeline operators

alcaeus committed:
Implement new aggregation pipeline operators in stages

alcaeus committed:
Update static analysis baselines

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -192,10 +278,62 @@ public function arrayElemAt($array, $index): self
+     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/bottom/
+     *
+     * @param mixed|self                $output
+     * @param array<string, int|string> $sortBy
```
Would it make sense to use a psalm type alias here for a sort spec?

alcaeus replied:
Yes, I'll re-use the sort types declared in #2513 once that's been merged.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -325,6 +463,175 @@ public static function convertExpression($expression)
+     *
+     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/count/
+     */
+    public function countDocuments(): self
```
Is this intentionally not named `count()` to avoid a potential conflict with Countable?

alcaeus replied:
The `count` accumulator is available in the `$group` stage for example, but every stage needs to declare methods for all other stages to allow for the fluent interface. This leads to a conflict between the `count` method to add a `$count` stage and the `$count` accumulator. In the stages, I've named the method `countDocuments`, and I decided to re-use that name for consistency here.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -325,6 +463,175 @@ public static function convertExpression($expression)
+     */
+    public function dateFromString($dateString, $format = null, $timezone = null, $onError = null, $onNull = null): self
+    {
+        return $this->operator(
```
Any reason this uses different formatting than `dateFromParts()` above?

alcaeus replied:
None other than me relying on phpcbf to format these. Since it doesn't apply consistent formatting for multi-line calls I've updated these manually to be consistent.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -325,6 +463,175 @@ public static function convertExpression($expression)
+    }
+
+    /**
+     * Decrements a Date object by a specified number of time units
```
Should "Date" be capitalized in this context? The casing is inconsistent across doc blocks, but if we're only talking about an object in the pipeline (i.e. BSON document with date fields within) I think lowercase makes sense to avoid anyone confusing this with a PHP object.

alcaeus replied:
While it may make sense in the MongoDB docs, there's no reason for this to be capitalised here. I've updated this throughout this file and the interfaces.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Operator/ArrayOperators.php:
```diff
@@ -0,0 +1,267 @@
+    public function first($expression): self;
+
+    /**
+     * Returns a specified number of elements from the beginning of an array. Distinct from the $firstN accumulator.
```
Should this be wrapped like other doc blocks?

alcaeus replied:
Wrapped description in interface docblocks at 80 chars.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -517,6 +853,39 @@ public function first($expression): self
     }
 
+    /**
+     * Returns a specified number of elements from the beginning of an array. Distinct from the $firstN accumulator.
```
Should this be wrapped like other doc blocks?

Also, it might be helpful to refer to "$group accumulator" for more context. I was initially confused because there are references to other accumulators in this class.

alcaeus replied:
Wrapped and updated to say "group accumulators" 👍

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -517,6 +853,39 @@ public function first($expression): self
+     *
+     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/function/
+     *
+     * @param string|Javascript $body
```
Is this actually enforced anywhere, or do you just rely on the server to do so?

alcaeus replied:
We rely on the server to validate this - mainly because the only reason we're not enforcing this (yet) is because we can't use a union type hint.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -538,6 +907,22 @@ public function getExpression(): array
     }
 
+    /**
+     * Returns the value of a specified field from a document. If you don't specify an object, $getField returns the value of the field from $$CURRENT.
```
Should this be wrapped? I'll stop mentioning this, as it's either intentional or you can address these on your own.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -903,10 +1304,38 @@ public function map($input, $as, $in): self
+     */
+    public function max($expression, ...$expressions): self
+    {
+        return $this->accumulatorOperator('$max', $expression, ...$expressions);
```
Alternatively, could you just splat `func_get_args()`? I assume this is slightly less overhead.

And the first argument is required because `$max` expects at least one expression, correct?

alcaeus replied:
Yes, I wanted to show that this takes at least one expression as an argument but can take more, hence the two parameters. I've also updated this to use `func_get_args()`.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -1110,6 +1564,16 @@ public function reverseArray($expression): self
+     */
+    public function sampleRate(float $rate): self
+    {
+        return $this->operator('$sampleRate', $rate);
```
I see that `operator()` calls `ensureArray()` on the second argument, which is a `float` in that case. Internally, it appears that `ensureArray()` will leave the `float` as-is, but that begs the question: should the method have a better name?

alcaeus replied:
Renamed the method to `prepareArgument` and documented it accordingly.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -1685,6 +2239,19 @@ private function operator(string $operator, $expression): self
+     */
+    public function or($expression, ...$expressions): self
+    {
+        return $this->operator('$or', func_get_args());
```
I noticed you used `func_get_args()` here but `$expression, ...$expressions` elsewhere. Should these be consistent?

alcaeus replied:
Simplified this for operators that take an array of arguments.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -1781,6 +2348,98 @@ public function objectToArray($object): self
     }
 
+    /**
+     * Provides regular expression (regex) pattern matching capability in
```
These methods are missing `@see` links.

alcaeus replied:
Added missing `@see` links to all methods in the interfaces.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -2059,4 +2710,23 @@ public function isNumber($expression): self
+             * @param mixed $value
+             * @param array-key $key
+             */
+            static fn ($value, $key): bool => ! in_array($key, $optionalArgNames) || $value !== null,
```
The `$value !== null` check is more efficient, so I'd lead with that to allow `in_array()` to be short-circuited.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Operator/AccumulatorOperators.php:
```diff
@@ -0,0 +1,89 @@
+ */
+interface AccumulatorOperators
+{
+    /**
```
I assume most of these doc blocks are duplicated in the interfaces and whatever implementations exist. Do you think they're prone to getting out of sync?

I assume you might need the `@param` stuff in both places for tooling.

alcaeus replied:
They are extremely prone to getting out of sync. I'll see if using `@inheritdoc` (or omitting the docblock entirely) works, or if we at least can omit the description and only keep type information.

alcaeus replied:
I played around with this and was able to remove the docblocks, keeping only the ones in the interfaces. Not only does this reduce the effort of keeping them in sync, but it also dramatically reduces the size of the `Expr` and `Operator` classes 🎉

jmikola reviewed (commented):
Not done with the review, but I'm publishing what comments I have so far.

I'm intentionally not reviewing the interfaces in detail and am instead focusing on the implementations.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Operator.php:
```diff
@@ -281,6 +326,51 @@ public function atanh($expression): self
+    /**
+     * Returns the average value of numeric values. Ignores non-numeric values.
+     *
+     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/avg/
```
Should this have `@see Expr::avg`? If so, you may want to check other doc blocks in this class for similar omissions.

alcaeus replied:
Added `@see` comments to all methods in the interfaces.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Operator.php:
```diff
@@ -281,6 +326,51 @@ public function atanh($expression): self
+     */
+    public function avg($expression, ...$expressions): self
+    {
+        $this->expr->avg(...func_get_args());
```
Is there a reason oyu use `...func_get_args()` here but `and()` above used `$expression, ...$expressions`? Also, is `func_get_args()`  problematic for psalm's type inference?

If `func_get_args()` is generally useful, I think it may be more beneficial in the date methods below, which have many more args.

alcaeus replied:
I've switched all methods to use `...func_get_args()`, regardless of the number of arguments. This makes it less error-prone when adding parameters or new methods (and more opportunity for copypasta)

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Operator.php:
```diff
@@ -1403,6 +1779,88 @@ public function reduce($input, $initialValue, $in): self
+     * Provides regular expression (regex) pattern matching capability in
+     * aggregation expressions.
+     *
+     * If a match is found, returns a document that contains information on the
```
The regex methods are missing `@see` annotations.

alcaeus replied:
Added to the interfaces.

jmikola commented on phpunit.xml.dist:
```diff
@@ -20,6 +20,10 @@
         </include>
+        <exclude>
+            <file>./lib/Doctrine/ODM/MongoDB/Aggregation/Stage/GraphLookup/Match.php</file>
+            <file>./lib/Doctrine/ODM/MongoDB/Aggregation/Stage/Match.php</file>
```
Why do these need to be excluded?

alcaeus replied:
When PHPUnit generates code coverage information, it trips over these due to the classes using a reserved identifier as class name.

jmikola replied:
Wait, isn't "match" [prohibited in PHP 8.0+](https://www.php.net/manual/en/reserved.keywords.php)? Ignoring PHPUnit, how is that allowed in the first place?

alcaeus replied:
All logic was moved from `Match` to `MatchStage`, and `Match` changed to extend `MatchStage`. Then, all ODM code was updated to only work with `MatchStage` instances. The `Match` classes are only loaded if a user specifically references them, which will cause errors on PHP 8.

On second thought, this introduced a subtle BC break which we didn't notice until now, but considering that nobody complained about it I'm not going to elaborate or care about it.

See https://github.com/doctrine/mongodb-odm/commit/8a5bf0e03f7c9be6b3f933559a353c912e0e8dfd for the commit that included the change.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/Aggregation/ExprTest.php:
```diff
@@ -18,7 +18,23 @@ class ExprTest extends BaseTest
      * @param Closure(Expr): mixed[]|mixed[] $args
      *
-     * @dataProvider provideAllOperators
+     * @dataProvider provideAccumulatorExpressionOperators
```
I like the idea of traits as data providers. This could be useful in PHPLIB to avoid stuffing all of our common providers in base classes.

alcaeus replied:
Yes, traits work great for composing functionality in PHPUnit tests to avoid duplicating tons of logic across test case classes. There are some more instances to clean up in ODM, but I figured this PR got unwieldy enough as it is.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/BucketAutoTest.php:
```diff
@@ -4,34 +4,46 @@
-    public function testBucketAutoStage(): void
+    /**
+     * @param array<string, string>          $expected
+     * @param Closure(Expr): mixed[]|mixed[] $args
```
Is the space after the colon here problematic for parsing the return type, or does the tooling allow for that?

alcaeus replied:
I've asked in the Symfony Slack's static-analysis channel as I don't have a good answer for this. I'll report back 👍

alcaeus replied:
The space is not problematic, but the union is. In this case I flipped the order of the types in the union, but in other cases it's a good idea to disambiguate this using parentheses.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/Aggregation/Stage/GroupTest.php:
```diff
@@ -17,57 +17,18 @@ class GroupTest extends BaseTest
-            'stdDevSamp()' => ['stdDevSamp', ['$field']],
-            'sum()' => ['sum', ['$field']],
-        ];
+        self::assertSame($groupStage, $groupStage->field('foo')->$operator(...$args));
```
Why does `BucketTest::testGroupAccumulators()` not perform a similar assertion on `$bucketStage`?

alcaeus replied:
Because once you get into the `output` of a `Bucket` stage, it no longer returns the stage itself but an output object for said stage. That makes it pretty much impossible to test.

alcaeus committed:
Use consistent formatting for operators with optional arguments

alcaeus committed:
Use consistent capitalisation for "date object"

alcaeus committed:
Wrap docblock comments

alcaeus committed:
Rename ensureArray to something more descriptive

alcaeus committed:
Remove duplicated docblocks

alcaeus committed:
Use func_get_args when proxying expression methods

alcaeus committed:
Fix wrong return types for internal expr methods

alcaeus committed:
Use func_get_args for operators where possible

alcaeus committed:
Disambiguate closure param in tests

alcaeus committed:
Fix docblock errors

Note: ProvidesGroupAccumulatorOperators errors were ignored as the trait is designed to be used in conjunction with the corresponding interface.

alcaeus commented:
Note: test failures on PHP 7.4 are due to `self` not being sufficient for what we want to do. Using a `static` return type will fix the problem but requires PHP 8. We'll most likely drop support for PHP 7.4 as a result.

alcaeus committed:
Use static return types across aggregation builder

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Expr.php:
```diff
@@ -584,205 +376,124 @@ public function dateFromString($dateString, $format = null, $timezone = null, $o
-        ], ['timezone', 'iso8601']));
+        return $this->operator(
+            '$dateToParts',
+            $this->filterOptionalNullArguments(
```
Just a thought but there may be an opporunity to replace the `$this->filterOptionalNullArguments()` stuff with another helper that builds an associative array from `func_get_args()` and filters automatically based on their default value, assuming this is common enough. I expect that would require method reflection so certainly a bit more overhead at runtime, though.

Not actually suggesting you pursue that, but I wanted to brain dump the idea.

alcaeus replied:
Yep, that would include some reflection logic as well, as `func_get_args` doesn't give us argument names. In a perfect world we'd use code generation for this, but introducing that in ODM is a big effort that I'm not sure is worth it.

jmikola commented on lib/Doctrine/ODM/MongoDB/Aggregation/Operator/CustomOperators.php:
```diff
@@ -28,15 +29,17 @@ interface CustomOperators
+     * @return static
      */
-    public function accumulator($init, $initArgs, $accumulate, $accumulateArgs, $merge, $finalize, $lang = 'js'): self;
+    public function accumulator($init, $initArgs, $accumulate, $accumulateArgs, $merge, $finalize = null, $lang = 'js'): self;
```
Good catch on `$finalize`. I realize `$initArgs` are also optional. Do you think it makes sense to default that to `null` as well? It'd be convenient for folks using named params, as they could omit it.

alcaeus replied:
Can do if we want to change the signature to have `$initArgs` just before `finalize`. Despite named arguments all parameters with a default value have to come after those without one. Happy to make the change if we don't care about keeping the arg list together with the code.

jmikola replied:
Ah, here's the deprecation notice: https://3v4l.org/QMnWc#v8.2.4

I think it's fine to move `$initArgs` to the first optional param. It's probably less likely to be used in practice -- assuming anyone even makes use of this helper to begin with.

alcaeus replied:
Done :)

jmikola reviewed (approved):
Feel free to make the change to `accumulator()` arg order on your own.

Everything else LGTM.

alcaeus committed:
Make initArgs parameter for $accumulator optional

alcaeus committed:
Fix handling of optional parameters for $range and $slice

alcaeus commented:
Ignoring PHPStan as those failures are the same we're already seeing due to psalm types.
